interfaces/apparmor/apparmor.go

```diff
@@ -0,0 +1,94 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package apparmor contains primitives for working with apparmor.
+//
+// References:
+//  - http://wiki.apparmor.net/index.php/Kernel_interfaces
+//  - http://apparmor.wiki.kernel.org/
+//  - http://manpages.ubuntu.com/manpages/xenial/en/man7/apparmor.7.html
+package apparmor
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// LoadProfile loads an apparmor profile from the given file.
+//
+// If no such profile was previously loaded then it is simply added to the kernel.
+// If there was a profile with the same name before, that profile is replaced.
+func LoadProfile(fname string) error {
+	return exec.Command("apparmor_parser", "--replace", fname).Run()
+}
+
+// Profile contains the name and mode of an apparmor profile loaded into the kernel.
+type Profile struct {
+	// Name of the profile. This is is either full path of the executable or an
+	// arbitrary string without spaces.
+	Name string
+	// Mode is either "enforcing" or "complaining".
+	Mode string
+}
+
+// Unload removes a profile from the running kernel.
+//
+// The operation is done with: apparmor_parser --remove $name
+func (profile *Profile) Unload() error {
+	return exec.Command("apparmor_parser", "--remove", profile.Name).Run()
+}
+
+// profilesPath contains information about the currently loaded apparmor profiles.
+const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
+
+var profilesPath = realProfilesPath
+
+// LoadedProfiles interrogates the kernel and returns a list of loaded apparmor profiles.
+//
+// Snappy manages apparmor profiles named *.snap. Other profiles might exist on
```

Comment:
On Fri, 2016-03-11 at 00:43 -0800, Zygmunt Krynicki wrote:
> > 
> > +
> > +// Unload removes a profile from the running kernel.
> > +//
> > +// The operation is done with: apparmor_parser --remove $name
> > +func (profile *Profile) Unload() error {
> > +	return exec.Command("apparmor_parser", "--remove",
> > profile.Name).Run()
> > +}
> > +
> > +// profilesPath contains information about the currently loaded
> > apparmor profiles.
> > +const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
> > +
> > +var profilesPath = realProfilesPath
> > +
> > +// LoadedProfiles interrogates the kernel and returns a list of
> > loaded apparmor profiles.
> > +//
> > +// Snappy manages apparmor profiles named *.snap. Other profiles
> > might exist on
> Why would the snap manager not use the rest api (and therefore the
> snappy api and all this code under the hood)? Ie, how can a snap
> manager *ever* remove snaps without snappy knowing it? Surely we
> aren't
> going to support a snap manager unmount and rm -f'ing things on its
> own?
> 
> ZK: The snap manager is part of the state engine. It lives in snapd
> itself (same process). Our design has a simple, fixed pipeline, with
> snap manager doing some stuff and then interface manager reacting as
> appropriate to change the security side of things.

Ok, good, that is what I figured and hoped, but I'm confused by this
statement then: "Since the snap manager can remove a snap without
telling us we need to notice that there are files and profiles in the
air that don't have a corresponding snap and act on them accordingly"

-- 
Jamie Strandboge             | http://www.canonical.com